elfo-network/src/socket/raw/tcp/mod.rs

```diff
@@ -0,0 +1,148 @@
+use std::net::SocketAddr;
+
+use derive_more::Display;
+use eyre::{Result, WrapErr};
+use futures::Stream;
+use tokio::net::{TcpListener, TcpStream};
+use tracing::warn;
+
+pub(super) use tokio::net::tcp::OwnedReadHalf;
+
+use std::{
+    io::{IoSlice, Result as IoResult},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::AsyncWrite;
+
+#[cfg(target_os = "linux")]
+mod metrics;
+
+#[cfg(target_os = "linux")]
+use self::metrics::TcpMetrics;
+
+pub(in crate::socket) struct OwnedWriteHalf {
+    inner: tokio::net::tcp::OwnedWriteHalf,
+    #[cfg(target_os = "linux")]
+    previous_metrics: TcpMetrics,
+}
+
+impl OwnedWriteHalf {
+    pub(crate) fn collect_transport_specific_metrics(&mut self) {
+        #[cfg(target_os = "linux")]
+        {
+            match metrics::collect_tcp_metrics(self.inner.as_ref()) {
+                Ok(current_metrics) => {
+                    metrics::report_tcp_metrics(&current_metrics, &self.previous_metrics);
+                    self.previous_metrics = current_metrics;
+                }
+                Err(err) => {
+                    tracing::warn!(
+                        message = "failed to collect TCP_INFO metrics",
```

Comment:
(minor)
strictly speaking, this level doesn't know that call is implemented using `TCP_INFO`, so it's leaky abstraction

Reply:
change TCP_INFO -> TCP